### ISSUE TITLE
HLS audio group for different quality videos (e.g. stereo for low quality video, surround for high quality video)

### ISSUE THREAD

**Issue by STiNG645** [google/shaka-packager#770]
HLS audio group for different quality videos (e.g. stereo for low quality video, surround for high quality video) 
Is audio grouping for video feature implemented? If yes, give an example please, I can not find (or understand) from the documentation.

**Comment by kqyang** [comment]
We do allow specification of audio groups through `--hls_group_id`: https://google.github.io/shaka-packager/html/tutorials/hls.html#hls-specific-stream-descriptor-fields.

But we generate variants for all audio groups. Assuming that there are two audio audios: stereo and surround, and there are two video streams: SD and HD, we will generate four StreamInfo entries (`#EXT-X-STREAM-INF`):

- SD with AUDIO="stereo"
- SD with AUDIO="surround"
- HD with AUDIO="stereo"
- HD with AUDIO="surround"

And you want a way to pick the audio groups and generate something like:

- SD with AUDIO="stereo"
- HD with AUDIO="surround"

Is my understanding correct?

**Comment by STiNG645** [comment]
Yes, right. I know about specification of audio groups through --hls_group_id, but I didn't find the function of attaching audio groups for different quality videos.

**Comment by kqyang** [comment]
This is not yet supported. 

It can be supported with a new stream descriptor like `hls_allowed_groups". Here is the change needed:

- Add a new stream descriptor similar to what we have done for `hls_group_id`: https://github.com/google/shaka-packager/blob/a1dd82d478c4622dff1a0d0489f170ed6e1814d5/packager/app/stream_descriptor.cc#L63

- Update corresponding documentations

- Propagate it to `MuxerListenerFactory` then to `HlsNotifyMuxerListener`: https://github.com/google/shaka-packager/blob/a1dd82d478c4622dff1a0d0489f170ed6e1814d5/packager/media/event/muxer_listener_factory.cc#L71

- Then to `MediaPlayList`: https://github.com/google/shaka-packager/blob/ab8fa87d1807e2f8a5756e365f9bf7b41cc4a992/packager/hls/base/simple_hls_notifier.cc#L277

- When generating StreamInfoTag, only allow groups that are specified: https://github.com/google/shaka-packager/blob/4f14a6f9739cca7b755d1dcaf5c93fb3e358f5a8/packager/hls/base/master_playlist.cc#L407

@STiNG645 You are welcomed to work on it if you are interested!

**Comment by STiNG645** [comment]
> This is not yet supported.
> 
> It can be supported with a new stream descriptor like `hls_allowed_groups". Here is the change needed:
> 
> * Add a new stream descriptor similar to what we have done for `hls_group_id`: https://github.com/google/shaka-packager/blob/a1dd82d478c4622dff1a0d0489f170ed6e1814d5/packager/app/stream_descriptor.cc#L63
> * Update corresponding documentations
> * Propagate it to `MuxerListenerFactory` then to `HlsNotifyMuxerListener`: https://github.com/google/shaka-packager/blob/a1dd82d478c4622dff1a0d0489f170ed6e1814d5/packager/media/event/muxer_listener_factory.cc#L71
> * Then to `MediaPlayList`: https://github.com/google/shaka-packager/blob/ab8fa87d1807e2f8a5756e365f9bf7b41cc4a992/packager/hls/base/simple_hls_notifier.cc#L277
> * When generating StreamInfoTag, only allow groups that are specified: https://github.com/google/shaka-packager/blob/4f14a6f9739cca7b755d1dcaf5c93fb3e358f5a8/packager/hls/base/master_playlist.cc#L407
> 
> @STiNG645 You are welcomed to work on it if you are interested!

Thanks for the answer, I try.
